src/orange/orangeinp/IntersectRegion.cc

```diff
@@ -334,6 +334,233 @@ void Ellipsoid::output(JsonPimpl* j) const
+{
+    for (auto i : range(2))
+    {
+        CELER_VALIDATE(radii_[i] >= 0, << "negative radius: " << radii_[i]);
+    }
+    CELER_VALIDATE(hh_ > 0, << "nonpositive halfheight: " << hh_);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Whether this encloses another elliptical cylinder.
+ */
+bool EllipticalCylinder::encloses(EllipticalCylinder const& other) const
+{
+    for (auto ax : range(2))
+    {
+        if (this->radii_[ax] < other.radii_[ax])
+        {
+            return false;
+        }
+    }
+    return hh_ >= other.hh_;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Build surfaces.
+ */
+void EllipticalCylinder::build(IntersectSurfaceBuilder& insert_surface) const
+{
+    insert_surface(Sense::outside, PlaneZ{-hh_});
+    insert_surface(Sense::inside, PlaneZ{hh_});
+
+    // Insert elliptical cylinder surface last, as a simple quadric with
+    // equation:
+    // r_y^2 x^2 + r_x^2 y^2 - r_x^2 r_y^2 = 0
+    real_type rx2 = ipow<2>(radii_[to_int(Axis::x)]);
+    real_type ry2 = ipow<2>(radii_[to_int(Axis::y)]);
+    real_type g = -rx2 * ry2;
+
+    Real3 abc{ry2, rx2, 0};
+    insert_surface(SimpleQuadric{abc, Real3{0, 0, 0}, g});
+
+    // Set exterior bbox
+    Real3 ex_halves{radii_[to_int(Axis::x)], radii_[to_int(Axis::y)], hh_};
+    insert_surface(Sense::inside, BBox{-ex_halves, ex_halves});
+
+    // Set an interior bbox (inscribed cuboid)
+    auto inv_sqrt_two = 1 / constants::sqrt_two;
+    Real3 in_halves{radii_[to_int(Axis::x)] * inv_sqrt_two,
+                    radii_[to_int(Axis::y)] * inv_sqrt_two,
+                    hh_};
+    insert_surface(Sense::outside, BBox{-in_halves, in_halves});
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Write output to the given JSON object.
+ */
+void EllipticalCylinder::output(JsonPimpl* j) const
+{
+    to_json_pimpl(j, *this);
+}
+
+//---------------------------------------------------------------------------//
+// ELLIPTICAL CONE
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with lower/upper x- and y-radii and half-height in z.
+ *
+ * The lower radii are the x- and y-radii at the plane z = -hh. The
+ * upper radii are the x- and y-radii at the plane z = hh. There are several
+ * restrictions on these radii:
+ *
+ * 1) Either the lower or upper radii may be (0, 0); this is the only permitted
+ *    way for the elliptical cone to include the vertex.
+ * 2) The aspect ratio of the elliptical cross sections is constant. Thus, the
+ *    aspect ratio at z = -hh must equal the aspect ratio at z = hh.
+ * 3) Degenerate elliptical cones with lower_radii == upper_radii (i.e.,
+ *    elliptical cylinders) are not permitted.
+ * 4) Degenerate elliptical cones where lower or upper radii are equal to
+ *    (0, x) or (x, 0), where x is non-zero, are not permitted.
```

Comment:
Since we generally don't show the documentation from constructors in the user doc, but this is good to have available on the ORANGE doc page, could you move this to the class documentation?